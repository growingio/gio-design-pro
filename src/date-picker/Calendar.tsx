import React, { useRef, useEffect } from 'react';
import zhCN from 'rc-calendar/lib/locale/zh_CN';
import RcDatePicker from 'rc-calendar/lib/Picker';
import RcRangeCalendar from 'rc-calendar/lib/RangeCalendar';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import moment, { Moment } from 'moment';
import { noop } from 'lodash';
import Button from '@gio-design/components/es/components/button';
import { Tabs, TabPane } from '@gio-design/components';
import Input from '@gio-design/components/es/components/input';
import { CalendarOutlined, PlusCircleFilled } from '@gio-design/icons';
import { DateRangePickerProps } from './interfaces';
import useCalendar from './hooks/useCalendar';
import './style';

moment.locale('zh-cn', {
  months: '1月_2月_3月_4月_5月_6月_7月_8月_9月_10月_11月_12月'.split('_'),
  monthsShort: '1月_2月_3月_4月_5月_6月_7月_8月_9月_10月_11月_12月'.split('_'),
  weekdaysMin: '日_一_二_三_四_五_六'.split('_'),
});

const disabledDate = (current: Moment) => {
  const date = moment();
  return current.isAfter(date);
};

const DateRangePicker: React.FC<DateRangePickerProps> = (props: DateRangePickerProps) => {
  const {
    prefixCls,
    format = 'YYYY/MM/DD',
    value,
    defaultValue,
    showFooter,
    // disabledDate,
    mode,
  } = props;

  const { state, actions } = useCalendar(props);

  const calendarContainerRef = useRef(null);

  const hackPanelClickToday = (e: any) => {
    if (e.target.className === 'gio-date-picker-date' && e.target.ariaDisabled === 'false') {
      const todayNode = document.querySelector('.gio-date-picker-today') as HTMLElement;
      todayNode.click();
    }
  };

  const hackPanelClickYesterday = (e: any) => {
    if (e.target.className === 'gio-date-picker-date' && e.target.ariaDisabled === 'false') {
      const yesterdayNode = document.querySelector('.gio-date-picker-today')?.previousSibling as HTMLElement;
      yesterdayNode.click();
    }
  };

  useEffect(() => {
    actions.setTimeRange(value);
  }, [value]);

  useEffect(() => {
    // eslint-disable-next-line react/no-find-dom-node
    const panelNode = ReactDOM.findDOMNode(calendarContainerRef.current) as HTMLElement;
    if (state.fixedMode) {
      // panelNode.addEventListener('click', handlePanelClick);
      if (state.endDay === 'yesterday') {
        panelNode.onclick = hackPanelClickYesterday;
      } else {
        panelNode.onclick = hackPanelClickToday;
      }
    } else {
      panelNode.onclick = noop;
    }
  }, [state.fixedMode, state.endDay]);

  useEffect(() => {
    if (mode === 'since') {
      actions.setTimeRange([value[0], moment()]);
      actions.setFixedMode(true);
    } else if (mode === 'dynamic') {
      actions.handleRightDynamicInput(7);
      actions.setFixedMode(true);
      actions.setLeftDynamicInputVisible(false);
      actions.setTimeRange([moment().subtract('days', 7), moment()]);
    } else {
      actions.setFixedMode(false);
      actions.setTimeRange(value);
    }
  }, [mode]);

  const CalendarCls = classNames(classNames, {
    [`${prefixCls}-no-footer`]: !showFooter,
  });

  const renderFooter = () => (
    <>
      {props.renderExtraFooter && (
        <div className={classNames(`${prefixCls}-extra-footer`)}>{props.renderExtraFooter()}</div>
      )}
      <Button onClick={actions.onCancel} type="secondary" size="middle" style={{ margin: ' 0 12px 0 0 ' }}>
        取消
      </Button>
      <Button onClick={actions.onConfirm} size="middle">
        确定
      </Button>
    </>
  );

  const formatDate = (v: Moment) => v?.format(format);

  const calendar = (
    <RcRangeCalendar
      locale={zhCN}
      format={format}
      defaultValue={defaultValue}
      disabledDate={disabledDate}
      // value={timeRange}
      onSelect={actions.onSelect}
      onPanelChange={actions.onPanelChange}
      showToday={false}
      showOk={false}
      showDateInput={false}
      prefixCls={prefixCls}
      className={CalendarCls}
      renderFooter={renderFooter}
    />
  );

  const renderPanel = (renderMode: string) => (
    <RcDatePicker
      calendar={calendar}
      value={state.timeRange}
      onChange={actions.onChange}
      prefixCls={`${prefixCls}-dropdown`}
      getCalendarContainer={() => calendarContainerRef.current}
      open={state.open}
    >
      {({ value: _value }: { value: Array<Moment> }) => (
        <div className={classNames(`${prefixCls}-range-input-${renderMode}`)}>
          {renderMode === 'absolute' && (
            <div className={classNames(`${prefixCls}-range-input`)}>
              <Input
                className={`${prefixCls}-input-first`}
                onChange={actions.handleLeftInputChange}
                value={state.leftInputTimeRange || `${formatDate(_value[0])}`}
                onClick={actions.handleOpen}
              />
              <span className={`${prefixCls}-split`}>—</span>
              <Input
                className={`${prefixCls}-input-second`}
                onChange={actions.handleRightInputChange}
                value={state.rightInputTimeRange || `${formatDate(_value[1])}`}
                onClick={actions.handleOpen}
                suffix={<CalendarOutlined />}
              />
            </div>
          )}
          {renderMode === 'since' && (
            <>
              从
              <Input
                className={`${prefixCls}-input-second`}
                onChange={actions.handleLeftInputChange}
                value={state.leftInputTimeRange || `${formatDate(_value[0])}`}
                onClick={actions.handleOpen}
                suffix={<CalendarOutlined />}
              />
              <Tabs
                activeKey={state.endDay}
                className={`${prefixCls}-tab`}
                size="middle"
                onChange={actions.handleEndDayChange}
              >
                <TabPane tab="至今日" key="today" style={{ height: '36px' }} />
                <TabPane tab="至昨日" key="yesterday" style={{ height: '36px' }} />
              </Tabs>
            </>
          )}
          {renderMode === 'dynamic' && !state.leftDynamicInputVisible && (
            <>
              过去
              <Input.InputNumber
                className={`${prefixCls}-input-second`}
                value={state.rightDynamicInput}
                onChange={actions.handleRightDynamicInput as any}
                min={0}
              />
              天
              <Button type="secondary" icon={<PlusCircleFilled />} onClick={actions.handleEndDay}>
                结束日期
              </Button>
            </>
          )}
          {renderMode === 'dynamic' && state.leftDynamicInputVisible && (
            <>
              过去
              <Input.InputNumber
                className={`${prefixCls}-input-second`}
                value={state.leftDynamicInput}
                onChange={actions.handleLeftDynamicInput}
                min={0}
              />
              至
              <Input.InputNumber
                className={`${prefixCls}-input-second`}
                value={state.rightDynamicInput}
                onChange={actions.handleRightDynamicInput as any}
                min={0}
              />
              天
            </>
          )}
          <div className={classNames(`${prefixCls}-wrapper`)} ref={calendarContainerRef} />
        </div>
      )}
    </RcDatePicker>
  );

  return <div className={classNames(`${prefixCls}-wrap-range`)}>{renderPanel(mode)}</div>;
};

export default DateRangePicker;
