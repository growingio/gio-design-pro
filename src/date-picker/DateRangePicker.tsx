/* eslint-disable @typescript-eslint/no-shadow */
import React from 'react';
import Button from '@gio-design/components/es/components/button';
import List from '@gio-design/components/es/components/list';
import Dropdown from '@gio-design/components/es/components/dropdown';
import usePrefixCls from '@gio-design/components/es/utils/hooks/use-prefix-cls';
import { CalendarOutlined } from '@gio-design/icons';
import useRangePicker from './hooks/useRangePicker';
import Calendar from './Calendar';
import { RangePickerProps } from './interfaces';

export enum Mode {
  shortcut = 'shortcut',
  since = 'since',
  dynamic = 'dynamic',
  absolute = 'absolute',
}

export const shortcutOptions = [
  [
    { value: 'day:1,0', label: '今日' },
    { value: 'week:1,0', label: '本周' },
    { value: 'month:1,0', label: '本月' },
    // { value: 'quarter:1,0', label: '本季度' },
    { value: 'year:1,0', label: '今年' },
    { value: 'day:8,1', label: '过去7天' },
    { value: 'day:31,1', label: '过去30天' },
    { value: 'day:181,1', label: '过去180天' },
    // { value: 'hour:24,0', label: '过去24小时' },
    { value: 'since', label: '开始至今' },
  ],
  [
    { value: 'day:2,1', label: '昨日' },
    { value: 'week:2,1', label: '上周' },
    { value: 'month:2,1', label: '上月' },
    // { value: 'quarter:2,1', label: '上季度' },
    { value: 'year:2,1', label: '去年' },
    { value: 'day:15,1', label: '过去14天' },
    { value: 'day:91,1', label: '过去90天' },
    { value: 'day:366,1', label: '过去365天' },
    // { value: 'hour:48,0', label: '过去48小时' },
  ],
];

const DateRangePicker: React.FC<RangePickerProps> = (props: RangePickerProps) => {
  const { prefixCls: customizePrefixCls } = props;
  // const [formatedTime, setFormatedTime] = useState(timeRange);
  const { state, actions, utils } = useRangePicker(props);

  const modeOptions = [
    { value: Mode.shortcut, label: '常用时间' },
    // { value: Mode.since, label: '自某天以后' },
    { value: Mode.dynamic, label: '过去动态时段' },
    { value: Mode.absolute, label: '过去固定时段' },
  ];

  // useEffect(() => {
  //   if (state.visible) {
  //     actions.setMode(Mode.shortcut);
  //   }
  // }, [state.visible]);

  const prefixCls = usePrefixCls('date-picker', customizePrefixCls);

  const renderCalendar = () => (
    <div
      style={{
        boxSizing: 'border-box',
        width: '540px',
        height: '430px',
        position: 'relative',
        display: 'block',
        lineHeight: 1.5,
        marginBottom: 16,
      }}
    >
      <Calendar
        prefixCls={prefixCls}
        value={state.time}
        onChange={actions.onChange}
        format={utils.format}
        onCancel={actions.onCancel}
        onConfirm={actions.onConfirm}
        mode={state.mode}
        showFooter
      />
    </div>
  );

  const renderShortcuts = () => (
    <div
      style={{
        display: 'inline-flex',
      }}
    >
      <List stateless dataSource={shortcutOptions[0]} width={128} onClick={actions.onListClick} />
      <List stateless dataSource={shortcutOptions[1]} width={128} onClick={actions.onListClick} />
    </div>
  );

  return (
    <Dropdown
      visible={state.visible}
      onVisibleChange={actions.setVisible}
      overlay={
        <div
          style={{
            width: 'auto',
            height: 'auto',
            backgroundColor: '#FFFFFF',
            display: 'inline-flex',
            justifyContent: 'center',
            borderRadius: '4px',
          }}
        >
          <List
            wrapStyle={{ borderRight: '1px solid #E7EAF9' }}
            dataSource={modeOptions}
            width={144}
            value={Mode.shortcut}
            onClick={actions.handleModeChange}
          />
          <div>{state.mode === Mode.shortcut ? renderShortcuts() : renderCalendar()}</div>
        </div>
      }
      placement="bottomRight"
    >
      <Button type="secondary" icon={<CalendarOutlined />}>
        {state.displayTime}
      </Button>
    </Dropdown>
  );
};

export default DateRangePicker;
