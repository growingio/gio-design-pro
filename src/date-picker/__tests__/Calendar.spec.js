import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount, render } from 'enzyme';
import moment from 'moment';
import { noop } from 'lodash';
import Calendar from '../Calendar';
import { Mode } from '../DateRangePicker';
// import Button from '../../button/button';

const format = 'YYYY/MM/DD';

const VALUE = [moment([2015, 5, 1]), moment([2015, 5, 2])];
// const disabledDate = () => false;

const props = {
  value: VALUE,
  onChange: noop,
  format: 'YYYY/MM/DD',
  onConfirm: noop,
  onCancel: noop,
  showFooter: true,
};

describe('DatePicker ui test', () => {
  const DatePickerInstance = () => <Calendar {...props} mode={Mode.absolute} />;

  it('should match snapshot', () => {
    const wrapper = render(DatePickerInstance());
    expect(wrapper).toMatchSnapshot();
  });

  //   const DatePickerInstance1 = () => <Calendar {...props} mode={Mode.shortcut} />;

  //   it('should match snapshot', () => {
  //     const wrapper = render(DatePickerInstance1());
  //     expect(wrapper).toMatchSnapshot();
  //   });

  //   const DatePickerInstance2 = () => <Calendar {...props} mode={Mode.since} />;

  //   it('should match snapshot', () => {
  //     const wrapper = render(DatePickerInstance2());
  //     expect(wrapper).toMatchSnapshot();
  //   });

  it('should receive props', () => {
    expect(() => {
      const wrapper = mount(DatePickerInstance());
      wrapper.setProps({ disabledDate: noop });
      wrapper.setProps({ onChange: noop });
      wrapper.setProps({ mode: Mode.absolute });
      wrapper.setProps({ onConfirm: noop });
      wrapper.setProps({ onCancel: noop });
      wrapper.setProps({ format });
      wrapper.setProps({ showFooter: 'true' });
      wrapper.unmount();
    }).not.toThrow();
  });
});
