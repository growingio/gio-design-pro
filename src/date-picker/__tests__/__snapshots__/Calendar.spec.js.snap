// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker ui test should match snapshot 1`] = `
<div
  class="undefined-wrap-range"
>
  <div
    class="undefined-range-input-absolute"
  >
    <div
      class="undefined-range-input"
    >
      <div
        class="gio-input undefined-input-first gio-input--middle"
      >
        <input
          class="gio-input__content"
          placeholder=""
          type="text"
          value="2015/06/01"
        />
      </div>
      <span
        class="undefined-split"
      >
        â€”
      </span>
      <div
        class="gio-input undefined-input-second gio-input--middle gio-input__container"
      >
        <input
          class="gio-input__content gio-input__content-suffix"
          placeholder=""
          type="text"
          value="2015/06/02"
        />
        <div
          class="gio-input__suffix"
        >
          <span
            aria-label="calendar-outlined"
            class="gio-icon"
            role="img"
          >
            <svg
              fill="currentColor"
              height="16px"
              viewBox="0 0 14.01 14"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs />
              <g>
                <g>
                  <path
                    d="M10.5 7h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm6 3h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3-6.95a.53.53 0 00-.5.53v7.93a.54.54 0 00.53.54h10.94a.54.54 0 00.53-.54V3.58a.53.53 0 00-.53-.53h-1.33a.13.13 0 00-.1 0 .12.12 0 000 .1v.35a.5.5 0 01-1 0v-.31a.14.14 0 00-.14-.14H4.14a.14.14 0 00-.14.14v.31a.5.5 0 01-1 0v-.31a.12.12 0 000-.1.13.13 0 00-.1 0zM9.86 2a.14.14 0 00.14-.14V1.5a.5.5 0 011 0v.36a.14.14 0 00.14.14H13a1 1 0 011 1v9a1 1 0 01-1 1H1a1 1 0 01-1-1V3a1 1 0 011-1h1.86A.14.14 0 003 1.86V1.5a.5.5 0 011 0v.36a.14.14 0 00.14.14z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    id="calendar-outlined_svg__calendar"
                  />
                </g>
              </g>
            </svg>
          </span>
        </div>
      </div>
    </div>
    <div
      class="undefined-wrapper"
    />
  </div>
</div>
`;
